lua/frecency/db.lua

```diff
@@ -0,0 +1,186 @@
+local util = require "frecency.util"
+local const = require "frecency.const"
+local algo = require "frecency.algo"
+
+local ok, sql = pcall(require, "sql")
+if not ok then
+  error "ERROR: frecency-telescope.nvim: sql.nvim (https://github.com/tami5/sql.nvim) is not found. Please install"
+  return
+end
+local s = sql.lib
+
+---@class FrecencyDB: SQLDatabaseExt
+---@field files SQLTableExt
+---@field timestamps SQLTableExt
+---@field workspaces SQLTableExt
+---@field config FrecencyConfig
+local db = sql {
+  uri = vim.fn.stdpath "data" .. "/file_frecency.sqlite3",
+  files = { id = true, count = { "integer", default = 0 }, path = "string" },
+  timestamps = { id = true, timestamp = "real", file_id = "integer" },
+}
+
+---@class FrecencyDBConfig
+---@field db_root string: default "${stdpath.data}/file_frecency.sqlite3"
+---@field ignore_patterns table: extra ignore patterns: default empty
+---@field safe_mode boolean: When enabled, the user will be prompted when entries > 10, default true
+---@field auto_validate boolean: When this to false, stale entries will never be automatically removed, default true
+db.config = {
+  db_root = nil,
+  ignore_patterns = {},
+  safe_mode = true,
+  auto_validate = true,
+}
+
+---Set database configuration
+---@param config FrecencyDBConfig
+db.set_config = function(config)
+  db.config = vim.tbl_extend("keep", config, db.config)
+end
+
+---Initialize frecency Database
+---@param config FrecencyDBConfig
+---@overload func()
+---@return FrecencyDB
+db.init = function(config)
+  config = config or {}
+  db.db.uri = config.db_root and config.db_root or db.db.uri
+  db.config = config and db.set_config(config) or db.config
+  db.db:init()
+  db.is_initialized = db.db.is_initialized ---TODO: remove when sql.nvim@#93 is fixed
+
+  ---Use oldfiles on first run.
+  if db.files:count() == 0 then
+    -- TODO: this needs to be scheduled for after shada load??
+    local oldfiles = vim.api.nvim_get_vvar "oldfiles"
+    for _, path in pairs(oldfiles) do
```

Comment:
i think ipairs is better here

Reply:
indeed < really how the f# could you miss that 🤣